src/emulator/modules/SceGxm/SceGxm.cpp

```diff
@@ -1010,27 +1010,64 @@ EXPORT(int, sceGxmSetFragmentTexture, SceGxmContext *context, unsigned int textu
 
     glActiveTexture((GLenum)(GL_TEXTURE0 + textureIndex));
     glBindTexture(GL_TEXTURE_2D, context->texture[0]);
+
+    // Disable mip-maps
     glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 0);
 
-    if (texture->format == SCE_GXM_TEXTURE_FORMAT_P8_ABGR) {
+    if (texture::is_paletted_format(texture->format)) {
+        const auto base_format = texture::get_base_format(texture->format);
+        const auto is_byte_indexed = (base_format == SCE_GXM_TEXTURE_BASE_FORMAT_P8); // only altenative is SCE_GXM_TEXTURE_BASE_FORMAT_P4
+        const auto palette_indexes = is_byte_indexed ? 256 : 16;
+
         glPixelTransferi(GL_MAP_COLOR, GL_TRUE);
+
+        // second dimension should be palette_indexes sized instead of 256, but prefer to keep it constant/stack-allocated
+        // TODO: Cache palette map instead of calculating here every time
         GLfloat map[4][256];
+
         const uint8_t(*const src)[4] = static_cast<uint8_t(*)[4]>(texture->palette.get(host.mem));
-        for (size_t i = 0; i < 256; ++i) {
+        for (size_t i = 0; i < palette_indexes; ++i) {
             map[0][i] = src[i][0] / 255.0f;
             map[1][i] = src[i][1] / 255.0f;
             map[2][i] = src[i][2] / 255.0f;
             map[3][i] = src[i][3] / 255.0f;
         }
-        glPixelMapfv(GL_PIXEL_MAP_I_TO_R, 256, map[0]);
-        glPixelMapfv(GL_PIXEL_MAP_I_TO_G, 256, map[1]);
-        glPixelMapfv(GL_PIXEL_MAP_I_TO_B, 256, map[2]);
-        glPixelMapfv(GL_PIXEL_MAP_I_TO_A, 256, map[3]);
+
+        // map channel indexes for each channel
+        // A == A_max means alpha channel is unused (max)
+        std::uint8_t R = 0, G = 0, B = 0, A = 0;
+        constexpr auto A_max = std::numeric_limits<decltype(A)>::max();
+
+        switch (texture::get_swizzle(texture->format)) {
+        case SCE_GXM_TEXTURE_SWIZZLE4_ABGR: R = 0; G = 1; B = 2; A = 3; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_ARGB: R = 2; G = 1; B = 0; A = 3; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_RGBA: R = 3; G = 2; B = 1; A = 0; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_BGRA: R = 1; G = 2; B = 3; A = 0; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_1BGR: R = 0; G = 1; B = 2; A = A_max; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_1RGB: R = 2; G = 1; B = 0; A = A_max; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_RGB1: R = 3; G = 2; B = 1; A = A_max; break;
+        case SCE_GXM_TEXTURE_SWIZZLE4_BGR1: R = 1; G = 2; B = 3; A = A_max; break;
+        default:
+        {
+            LOG_ERROR("Invalid swizzle for paletted texture foramt.");
+        }
+        }
+
+        glPixelMapfv(GL_PIXEL_MAP_I_TO_R, palette_indexes, map[R]);
+        glPixelMapfv(GL_PIXEL_MAP_I_TO_G, palette_indexes, map[G]);
+        glPixelMapfv(GL_PIXEL_MAP_I_TO_B, palette_indexes, map[B]);
+        if (A == A_max) {
+            constexpr float max_alpha[1] = { 255.0 };
```

Comment:
Is this array big enough? Should it be 256 elements?

Reply:
Ah indeed it should, gonna fix that later.

Reply:
Pushed a fix.
Turns out constexpr float array initialization is impossible without really hacky stuff, so I just used `std::fill`.